functions/go/apply-setters/set.go

```diff
@@ -0,0 +1,223 @@
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+var _ kio.Filter = Set{}
+
+// Set sets resource field values with setter references
+type Set struct {
+	// Setters holds the user provided values for all the setters
+	Setters []Setter `json:"setters,omitempty" yaml:"setters,omitempty"`
+}
+
+type Setter struct {
+	// Name is the name of the setter
+	Name string `json:"name,omitempty" yaml:"name,omitempty"`
+
+	// Value is the input value for setter
+	Value string `json:"value,omitempty" yaml:"value,omitempty"`
+}
+
+const SetPrefix = "# kpt-set: "
+
+// Filter implements Set as a yaml.Filter
+func (s Set) Filter(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	for i := range nodes {
+		err := accept(&s, nodes[i])
+		if err != nil {
+			return nil, errors.Wrap(err)
+		}
+	}
+	return nodes, nil
+}
+
+// visitMapping
+func (s *Set) visitMapping(object *yaml.RNode, path string) error {
+	return object.VisitFields(func(node *yaml.MapNode) error {
+		if node == nil || node.Key == nil || node.Key.YNode() == nil ||
+			node.Value == nil || node.Value.YNode() == nil {
+			return nil
+		}
+		// the aim of this method is to apply-setter for sequence nodes
+		if node.Value.YNode().Kind != yaml.SequenceNode {
+			// return if it is not a sequence node
+			return nil
+		}
+
+		setterComment := node.Key.YNode().LineComment
+		if !strings.HasPrefix(setterComment, SetPrefix) {
+			return nil
+		}
+		setterComment = strings.TrimSpace(strings.TrimPrefix(setterComment, SetPrefix))
+		if !shouldSet(setterComment, s.Setters) {
+			// this means there is no intent from user to modify this setter tagged resources
+			return nil
+		}
+
+		// since this setter pattern is found on sequence node, make sure that it is
+		// not interpolation of setters, it should be simple setter e.g. ${environments}
+		if !validArraySetterPattern(setterComment) {
+			return errors.Errorf("setter interpolation is not allowed for sequence nodes: %q", setterComment)
+		}
+
+		// get the setter value for the setter name in the comment
+		sv := setterValue(s.Setters, setterComment)
+
+		// parse the setter value as yaml node
+		rn, err := yaml.Parse(sv)
+		if err != nil {
+			return err
+		}
+
+		// the setter value must parse as sequence node
+		if rn.YNode().Kind != yaml.SequenceNode {
+			return errors.Errorf("input to array setter must parse into sequence node, but found %q", sv)
+		}
+
+		node.Value.YNode().Content = rn.YNode().Content
+		node.Value.YNode().Style = yaml.FoldedStyle
+		return nil
+	})
+}
+
+// visitSequence
+func (s *Set) visitSequence(_ *yaml.RNode, _ string) error {
+	// this need not be implemented
+	return nil
+}
+
+// visitScalar
+func (s *Set) visitScalar(object *yaml.RNode, _ string) error {
+	if object == nil || object.YNode() == nil {
+		return nil
+	}
+	// perform a direct set of the field if it matches
+	setterComment := object.YNode().LineComment
+	if !strings.HasPrefix(setterComment, SetPrefix) {
```

Comment:
Extract functions to improve readability of the methods here.

e.g hasSetterReference()